Expose entities for Schneider Electric switches/dimmers, add more models

uvNikita opened the pull request:
## Proposed change

Supersedes: https://github.com/zigpy/zha-device-handlers/pull/3615
Fixes: https://github.com/zigpy/zha-device-handlers/issues/2828, fixes https://github.com/zigpy/zha-device-handlers/issues/2798, fixes https://github.com/zigpy/zha-device-handlers/issues/1696, fixes https://github.com/zigpy/zha-device-handlers/issues/4409
Part of: https://github.com/zigpy/zha-device-handlers/pull/1705

## Additional information
Spec docs:
[ZB Spec - Micro Module Dimmer -110422.pdf](https://github.com/user-attachments/files/22274736/ZB.Spec.-.Micro.Module.Dimmer.-110422.pdf)
[ZB Spec - Micro Module Switch - 110422.pdf](https://github.com/user-attachments/files/22274738/ZB.Spec.-.Micro.Module.Switch.-.110422.pdf)
[ZB Spec - Relay Switch 10A - 110422.pdf](https://github.com/user-attachments/files/22274739/ZB.Spec.-.Relay.Switch.10A.-.110422.pdf)
[ZB Spec - Rotary Dimmer - 110422.pdf](https://github.com/user-attachments/files/22274741/ZB.Spec.-.Rotary.Dimmer.-.110422.pdf)

Switches:
<img width="300" height="1408" alt="image" src="https://github.com/user-attachments/assets/69ab60c6-77fb-41c9-b146-446d55cd3608" />

Dimmers:
<img width="300" height="2044" alt="image" src="https://github.com/user-attachments/assets/be49e11b-63ea-4eca-bb40-a769cbc1ad6c" />



## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works
- [ ] Device diagnostics data has been attached

uvNikita committed:
Expose entities for Schneider Electric switches/dimmers, add more models

uvNikita committed:
Expose entities for Schneider Electric switches/dimmers, add more models

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

Orange-GT3 commented:
Amazing work, thanks so much @uvNikita.

TheJulianJES commented on zhaquirks/schneiderelectric/dimmers.py:
```diff
@@ -1,35 +1,143 @@
+    )
 )
 
+(BASE_DIMMER.clone().applies_to(SE_MANUF_NAME, "PUCK/DIMMER/1").add_to_registry())
```
You can use `# fmt: skip` something like this to format this in a nice way:
```suggestion
(
    BASE_DIMMER.clone()
    .applies_to(SE_MANUF_NAME, "PUCK/DIMMER/1")
    .add_to_registry()
)  # fmt: skip
```

TheJulianJES commented on zhaquirks/schneiderelectric/dimmers.py:
```diff
@@ -1,35 +1,143 @@
-    .applies_to(SE_MANUF_NAME, "NHROTARY/UNIDIM/1")
-    .applies_to(SE_MANUF_NAME, "NHPB/DIMMER/1")
-    .applies_to(SE_MANUF_NAME, "NHPB/UNIDIM/1")
+BASE_DIMMER = (
```
I think we should just use snake case: `base_dimmer`

uvNikita committed:
Lowercase and renames

- BASE_DIMMER -> base_micro_dimmer
- BASE_ROTARY_DIMMER -> base_dimmer
- BASE_SWITCH -> base_micro_switch

uvNikita committed:
Lowercase and renames

- BASE_DIMMER -> base_micro_dimmer
- BASE_ROTARY_DIMMER -> base_dimmer
- BASE_SWITCH -> base_micro_switch

uvNikita commented:
@TheJulianJES  implemented suggested changes.

I also renamed base_dimmer to base_micro_dimmer and base_rotary_dimmer to base_dimmer.

These quirks cover micro modules that go behind the existing switch (these devices have the base clusters/attributes), full switches/dimmers with push buttons, and rotary dimmers.

nnae06 commented:
Hello.

My entire house consists of various Schneider products. Both the switch and dimmer functions work perfectly.

I would also like to add Schneider motion sensors, I am missing some functions.

NHMOTION/UNIDIM/1 (WDE002962)

Maybe it would have been possible to add the functions.
- Fade time
- Sensitivity setting
- Detection delay
- Lux adjustment

nnae06 commented:
Can also confirm that the functions work together with Merten / Schneider switches and dimmers.

Have tried.
Switch: 2GANG/SWITCH/2
Dimmer: 1GANG/DIMMER/1

uvNikita commented:
@TheJulianJES I think this PR should be ready to merge. Let me know if there is anything missing.

Orange-GT3 commented:
I have some 10AX switches now so would love to see this PR merged: @TheJulianJES?

nnae06 commented:
I have now been running this code for a few months without any problems. Everything works flawlessly.

"""Schneider Electric dimmers quirks."""

from zigpy.quirks.v2 import EntityPlatform, EntityType, QuirkBuilder
from zigpy.quirks.v2.homeassistant import UnitOfTime
from zigpy.quirks.v2.homeassistant.number import NumberDeviceClass

from zhaquirks.schneiderelectric import (
    SE_MANUF_NAME,
    SEBallast,
    SEBasic,
    SEControlMode,
    SEDimmingCurve,
    SEOnOff,
    SEOnTimeReloadOptions,
    SESwitchAction,
    SESwitchConfiguration,
    SESwitchIndication,
    SEWiringMode,
)

base_micro_dimmer = (
    QuirkBuilder()
    .replaces(SEBasic, endpoint_id=3)
    .replaces(SEBallast, endpoint_id=3)
    .replaces(SEOnOff, endpoint_id=3)
    .number(
        attribute_name=SEBallast.AttributeDefs.min_level.name,
        cluster_id=SEBallast.cluster_id,
        endpoint_id=3,
        min_value=1,
        max_value=254,
        step=1,
        fallback_name="Min light level",
        translation_key="min_level",
    )
    .number(
        attribute_name=SEBallast.AttributeDefs.max_level.name,
        cluster_id=SEBallast.cluster_id,
        endpoint_id=3,
        min_value=1,
        max_value=254,
        step=1,
        fallback_name="Max light level",
        translation_key="max_level",
    )
    .number(
        attribute_name=SEOnOff.AttributeDefs.se_on_time_reload.name,
        cluster_id=SEOnOff.cluster_id,
        endpoint_id=3,
        min_value=0x0,
        max_value=0xFFFFFFFF,
        step=1,
        unit=UnitOfTime.SECONDS,
        device_class=NumberDeviceClass.DURATION,
        translation_key="on_time_reload",
        fallback_name="On time reload",
    )
    .number(
        attribute_name=SEOnOff.AttributeDefs.se_pre_warning_time.name,
        cluster_id=SEOnOff.cluster_id,
        endpoint_id=3,
        min_value=0x0,
        max_value=0xFFFF,
        step=1,
        unit=UnitOfTime.SECONDS,
        device_class=NumberDeviceClass.DURATION,
        translation_key="pre_warning_time",
        fallback_name="Pre warning time",
    )
    .enum(
        endpoint_id=3,
        cluster_id=SEOnOff.cluster_id,
        attribute_name=SEOnOff.AttributeDefs.se_on_time_reload_options.name,
        enum_class=SEOnTimeReloadOptions,
        translation_key="on_time_reload_options",
        fallback_name="On time reload options",
    )
    .enum(
        endpoint_id=3,
        cluster_id=SEBallast.cluster_id,
        attribute_name=SEBallast.AttributeDefs.se_control_mode.name,
        enum_class=SEControlMode,
        translation_key="control_mode",
        fallback_name="Control mode",
    )
)

base_dimmer = (
    base_micro_dimmer.clone()
    .replaces(SEBasic, endpoint_id=21)
    .replaces(SESwitchConfiguration, endpoint_id=21)
    .enum(
        endpoint_id=21,
        cluster_id=SESwitchConfiguration.cluster_id,
        attribute_name=SESwitchConfiguration.AttributeDefs.se_switch_indication.name,
        enum_class=SESwitchIndication,
        translation_key="switch_indication",
        fallback_name="Switch indication",
    )
    .enum(
        endpoint_id=21,
        cluster_id=SESwitchConfiguration.cluster_id,
        attribute_name=SESwitchConfiguration.AttributeDefs.se_switch_actions.name,
        enum_class=SESwitchAction,
        translation_key="switch_actions",
        fallback_name="Switch actions",
    )
)

(
    base_micro_dimmer.clone()
    .applies_to(SE_MANUF_NAME, "PUCK/DIMMER/1")
    .add_to_registry()
)  # fmt: skip

(
    base_dimmer.clone()
    .applies_to(SE_MANUF_NAME, "NHROTARY/DIMMER/1")
    .applies_to(SE_MANUF_NAME, "NHPB/DIMMER/1")
    .applies_to(SE_MANUF_NAME, "CH/DIMMER/1")
    .applies_to(SE_MANUF_NAME, "1GANG/DIMMER/1")
    .add_to_registry()
)

(
    base_dimmer.clone()
    .applies_to(SE_MANUF_NAME, "NHROTARY/UNIDIM/1")
    .applies_to(SE_MANUF_NAME, "NHPB/UNIDIM/1")
    .enum(
        endpoint_id=3,
        cluster_id=SEBallast.cluster_id,
        attribute_name=SEBallast.AttributeDefs.se_wiring_mode.name,
        enum_class=SEWiringMode,
        entity_platform=EntityPlatform.SENSOR,
        entity_type=EntityType.DIAGNOSTIC,
        translation_key="wiring_mode",
        fallback_name="Wiring mode",
    )
    .enum(
        endpoint_id=3,
        cluster_id=SEBallast.cluster_id,
        attribute_name=SEBallast.AttributeDefs.se_dimming_curve.name,
        enum_class=SEDimmingCurve,
        translation_key="dimming_curve",
        fallback_name="Dimming curve",
    )
    .add_to_registry()
)


-----------------------------------------------------------------------------------------------------------------------------------------------


"""Schneider Electric switches quirks."""

from zigpy.quirks.v2 import QuirkBuilder
from zigpy.quirks.v2.homeassistant import UnitOfTime
from zigpy.quirks.v2.homeassistant.number import NumberDeviceClass

from zhaquirks.schneiderelectric import (
    SE_MANUF_NAME,
    SEBasic,
    SEOnOff,
    SEOnTimeReloadOptions,
    SESwitchAction,
    SESwitchConfiguration,
    SESwitchIndication,
)

base_micro_switch = (
    QuirkBuilder()
    .replaces(SEBasic)
    .replaces(SEOnOff)
    .number(
        attribute_name=SEOnOff.AttributeDefs.se_on_time_reload.name,
        cluster_id=SEOnOff.cluster_id,
        endpoint_id=1,
        min_value=0x0,
        max_value=0xFFFFFFFF,
        step=1,
        unit=UnitOfTime.SECONDS,
        device_class=NumberDeviceClass.DURATION,
        translation_key="on_time_reload",
        fallback_name="On time reload",
    )
    .number(
        attribute_name=SEOnOff.AttributeDefs.se_pre_warning_time.name,
        cluster_id=SEOnOff.cluster_id,
        endpoint_id=1,
        min_value=0x0,
        max_value=0xFFFF,
        step=1,
        unit=UnitOfTime.SECONDS,
        device_class=NumberDeviceClass.DURATION,
        translation_key="pre_warning_time",
        fallback_name="Pre warning time",
    )
    .enum(
        endpoint_id=1,
        cluster_id=SEOnOff.cluster_id,
        attribute_name=SEOnOff.AttributeDefs.se_on_time_reload_options.name,
        enum_class=SEOnTimeReloadOptions,
        translation_key="on_time_reload_options",
        fallback_name="On time reload options",
    )
)

(
    base_micro_switch.clone()
    .applies_to(SE_MANUF_NAME, "NHPB/SWITCH/1")
    .applies_to(SE_MANUF_NAME, "2GANG/SWITCH/2")
    .applies_to(SE_MANUF_NAME, "CH2AX/SWITCH/1")
    .applies_to(SE_MANUF_NAME, "CH10AX/SWITCH/1")
    .replaces(SEBasic, endpoint_id=21)
    .replaces(SESwitchConfiguration, endpoint_id=21)
    .enum(
        endpoint_id=21,
        cluster_id=SESwitchConfiguration.cluster_id,
        attribute_name=SESwitchConfiguration.AttributeDefs.se_switch_indication.name,
        enum_class=SESwitchIndication,
        translation_key="switch_indication",
        fallback_name="Switch indication",
    )
    .enum(
        endpoint_id=21,
        cluster_id=SESwitchConfiguration.cluster_id,
        attribute_name=SESwitchConfiguration.AttributeDefs.se_switch_actions.name,
        enum_class=SESwitchAction,
        translation_key="switch_actions",
        fallback_name="Switch actions",
    )
    .add_to_registry()
)

(
    base_micro_switch.clone()
    .applies_to(SE_MANUF_NAME, "PUCK/SWITCH/1")
    .add_to_registry()
)  # fmt: skip

Orange-GT3 commented:
> I have now been running this code for a few months without any problems. Everything works flawlessly.

Great to hear. Can you edit your post to put the code blocks inside these tags ```? Three at the start and three at the end. This preserves the correct formatting and indentation for the code.

For example:
```
from zigpy import types as t
from zigpy.quirks.v2 import QuirkBuilder, ReportingConfig
from zigpy.quirks.v2.homeassistant import (
    PERCENTAGE,
    EntityType,
    UnitOfElectricPotential,
    UnitOfLength,
    UnitOfTemperature,
    UnitOfTime,
)
```

nnae06 commented:
I'll fix it a little later in the day.

nnae06 commented:
[dimmers.py](https://github.com/user-attachments/files/26546722/dimmers.py)
[switches.py](https://github.com/user-attachments/files/26546723/switches.py)

Orange-GT3 commented:
> [dimmers.py](https://github.com/user-attachments/files/26546722/dimmers.py) [switches.py](https://github.com/user-attachments/files/26546723/switches.py)

That's great, thank you. I have loaded the switches quirk and it seems to be working:

<img width="437" height="530" alt="image" src="https://github.com/user-attachments/assets/d8eb3e4c-e23b-4661-b3c2-8be4f135e0ad" />

It's good to be able to set the `Switch indication` through the UI, although I have no idea what the first 4 settings do and there are are a surpisingly large number of options for `Switch actions`:
<img width="358" height="720" alt="image" src="https://github.com/user-attachments/assets/68d6d8ce-1fcb-4c28-891a-f1f039287c24" />

I'm not going to fiddle, in case I break something, but do you know what these settings are used for?

uvNikita commented:
Rebased on top of the current dev. @TheJulianJES I think it should be ready for merge.

uvNikita committed:
Fix SE dimmer/switch entity ranges and styling

TheJulianJES commented on zhaquirks/schneiderelectric/dimmers.py:
```diff
@@ -1,34 +1,150 @@
     .replaces(SEBallast, endpoint_id=3)
     .replaces(SEOnOff, endpoint_id=3)
+    .number(
+        attribute_name=SEBallast.AttributeDefs.min_level.name,
```
Just a note, these were added to ZHA with:
- https://github.com/zigpy/zha/pull/659

uvNikita replied:
Great, I removed duplicated declarations now

TheJulianJES commented on zhaquirks/schneiderelectric/dimmers.py:
```diff
@@ -1,34 +1,150 @@
+        fallback_name="Pre warning time",
+    )
+    .enum(
+        attribute_name=SEOnOff.AttributeDefs.se_on_time_reload_options.name,
```
Writing bitmaps is a annoying. There is a branch I'll need to update which (IIRC) would add some quirks v2 builder methods for switches, with the "magic" handled behind the scenes.

I think switches would work here then? If we want to merge this PR next, we could also remove this temporarily. Smaller PRs are always easier to merge but I'm sorry about the wait on all your PRs. We're cleaning up the backlog next and are through with some big ZHA refactors, so hopefully we'll have speedier reviews soon.

uvNikita replied:
In the spec doc, they make it sound like the two options are mutually exclusive, that's why I implemented it as a select originally:

<img width="1254" height="158" alt="image" src="https://github.com/user-attachments/assets/eefd5d6e-7bb9-4a49-b9ba-d5da100abb44" />

But I removed the select definition for now until we have a better way of handling the bitmaps.

uvNikita committed:
Address review feedback for SE switches/dimmers

- Remove min_level/max_level numbers
- Remove se_on_time_reload_options bitmap-as-SELECT entities

uvNikita commented:
- Removed the duplicate min/max level numbers
- Removed se_on_time_reload_options select until we have better way to handle bitmaps
- Rebased on top of the current dev

uvNikita commented on zhaquirks/schneiderelectric/dimmers.py:
```diff
@@ -1,34 +1,150 @@
+        fallback_name="Wiring mode",
+    )
+    .enum(
+        attribute_name=SEBallast.AttributeDefs.se_dimming_curve.name,
```
At least my firmware doesn't reject the write to this attribute. I'm not sure if it changes anything though.
I kept it since it's part of the spec and writes succeed.
